index.js

```diff
@@ -34,19 +38,201 @@ class ServerlessPlugin {
             }
         }
     }
+
+    async createVendorZip() {
+        if(! this.serverless.service.custom.separateVendor) {
+            return;
+        }
+
+        this.bucketName = await this.provider.getServerlessDeploymentBucketName();
+        this.deploymentPrefix = await this.provider.getDeploymentPrefix();
+
+        const vendorZipHash = await this.createZipFile();
+        const newVendorZipName = vendorZipHash + '.zip';
+
+        this.fs.renameSync('vendor.zip', newVendorZipName);
+
+        await this.uploadZipToS3(newVendorZipName);
+
+        console.log('Bref: Setting environment variables.');
+
+        let filePath = this.stripSlashes((this.deploymentPrefix || '') + '/vendors/' + newVendorZipName);
+
+        let excludes = this.serverless.service.package.exclude;
+        if(excludes.indexOf('vendor/**') === -1) {
+            excludes[excludes.length] = 'vendor/**';
+        }
+
+        excludes[excludes.length] = newVendorZipName;
+
+        this.serverless.service.provider.environment.BREF_DOWNLOAD_VENDOR = `s3://${this.bucketName}/${filePath}`;
+
+        let iamRoleStatements = this.serverless.service.provider.iamRoleStatements;
+        const roleDetails = {
+            'Effect': 'Allow',
+            'Action': 's3:GetObject',
+            'Resource': [
+                `${this.bucketName}/vendors/*`
+            ]
+        };
+
+        if(typeof iamRoleStatements !== 'undefined' && iamRoleStatements) {
+            if(iamRoleStatements.indexOf(roleDetails) === -1) {
+                iamRoleStatements[iamRoleStatements.length] = roleDetails;
+            }
+        } else {
+            this.serverless.service.provider.iamRoleStatements = [
+                roleDetails
+            ];
+        }
+
+        console.log(this.serverless.service.provider.environment);
+
+        console.log('Bref: Vendor separation done!');
+    }
+
+    async createZipFile() {
+        const filePath = 'vendor.zip';
+
+        return await new Promise((resolve, reject) => {
+            const archiver = require(process.mainModule.path + '/../node_modules/archiver');
+            const output = this.fs.createWriteStream(filePath);
+            const archive = archiver('zip', {
+                zlib: { level: 9 } // Sets the compression level.
+            });
+
+            console.log(`Bref: Creating ${filePath} archive...`);
+
+            archive.pipe(output);
+            archive.directory('vendor/', false);
+            archive.finalize();
+
+            output.on('close', () => {
+                console.log(`Bref: Created ${filePath} with ${archive.pointer()} total bytes.`);
+                resolve();
+            });
+
+            output.on('end', () => {
+                console.log('Bref: Archiver data stream has been drained');
```

Comment:
Is this a good thing or a bad thing? I'd like to improve this message if it's an error, if it may duplicate the other message above?

Reply:
Removed 😃